src/Entities/Pawn.cpp

```diff
@@ -111,6 +111,44 @@ void cPawn::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
 void cPawn::KilledBy(TakeDamageInfo & a_TDI)
 {
 	ClearEntityEffects();
+
+	if (IsPlayer())
+	{
+		auto Player = static_cast<cPlayer *>(this);
+		cInventory & inv = Player->GetInventory();
+
+		//  If the player is holding a totem of undying in their off-hand or
+		//  main-hand slot and receives otherwise fatal damage, the totem saves the player from death.
+		if ((inv.GetShieldSlot().m_ItemType == E_ITEM_TOTEM_OF_UNDYING ||
+			 inv.GetEquippedItem().m_ItemType == E_ITEM_TOTEM_OF_UNDYING) &&
+				a_TDI.DamageType != dtAdmin && a_TDI.DamageType != dtInVoid
+			)
+		{
+			if (inv.GetShieldSlot().m_ItemType == E_ITEM_TOTEM_OF_UNDYING) {
+				inv.SetShieldSlot(cItem());
+			}
+			else
+			{
+				inv.SetEquippedItem(cItem());
+			}
```

Comment:
It's a bit ugly checking the equipped item in the previous if-statement, then immediately checking it again, how about extracting this into a seperate function?

```suggestion
		// Is death eligible for totem reanimation?
		if (a_TDI.DamageType != dtAdmin && a_TDI.DamageType != dtInVoid && DeductTotem(Player))
		{
```

where DeductTotem is implemented similarly:
```cpp
bool DeductTotem(cPlayer * a_Player)
{
	auto & inv = a_Player->GetInventory();
	if (inv.GetShieldSlot().m_ItemType == E_ITEM_TOTEM_OF_UNDYING)
	{
		inv.SetShieldSlot(cItem());
		return true;
	}
	if (inv.GetEquippedItem().m_ItemType == E_ITEM_TOTEM_OF_UNDYING)
	{
		inv.SetEquippedItem(cItem());
		return true;
	}
	return false;
}
```

Reply:
Actually, there's another problem here too. The equipped totem should be removed before the shield slot totem - I've fixed my suggestion above but yours also has that issue.

Reply:
I made the change you suggested, it works but travis ci builds failed with 2 errors. Second error is absurd and first weird.
`/home/travis/build/cuberite/cuberite/src/Entities/Pawn.cpp:109:6: error: no previous prototype for function 'DeductTotem' [-Werror,-Wmissing-prototypes]
bool DeductTotem(cPlayer * a_Player)
     ^
/home/travis/build/cuberite/cuberite/src/Entities/Pawn.cpp:133:16: error: unused variable 'inv' [-Werror,-Wunused-variable]
                cInventory & inv = Player->GetInventory();
                             ^
2 errors generated.`

Reply:
The first error can be resolved by adding a declaration for DeductToken to Pawn.h or (preferably) making the function static. 

The second is simple, you just need to remove the definition of inv on line 133, as it is no longer used (now it's part of the function you moved the rest into).